operator/controllers/loki/lokistack_controller.go

```diff
@@ -329,3 +333,58 @@ func (r *LokiStackReconciler) enqueueForStorageSecret() handler.EventHandler {
 		return requests
 	})
 }
+
+func (r *LokiStackReconciler) enqueueForNodeLabelChanges() handler.EventHandler {
+	ctx := context.TODO()
+	return handler.EnqueueRequestsFromMapFunc(func(obj client.Object) []reconcile.Request {
+		lokiStacks := &lokiv1.LokiStackList{}
+		if err := r.Client.List(ctx, lokiStacks); err != nil {
+			r.Log.Error(err, "Error getting LokiStack resources in event handler")
+			return nil
+		}
+
+		var requests []reconcile.Request
+		for _, stack := range lokiStacks.Items {
+			podList := &corev1.PodList{}
+			if err := r.Client.List(ctx, podList, &client.ListOptions{Namespace: stack.Namespace}); err != nil {
+				r.Log.Error(err, "Error getting pod resources in event handler")
+				return nil
+			}
+
+			for _, pod := range podList.Items {
+				assignedNode := &corev1.Node{}
+				key := client.ObjectKey{Name: pod.Spec.NodeName}
+				if err := r.Client.Get(ctx, key, assignedNode); err != nil {
+					r.Log.Error(err, "Error getting node resources in event handler")
+					return nil
+				}
+
+				labelsAnnotation, ok := pod.Annotations[lokiv1.AnnotationAvailabilityZoneLabels]
+				if !ok {
+					err := kverrors.New("Zone-aware pod is missing node-labels annotation", "annotation", lokiv1.AnnotationAvailabilityZoneLabels)
+					r.Log.Error(err, "Zone-aware pod is missing node-labels annotation")
+					return nil
+				}
+				nodeLabels := generateTopologyKeyAnnotationFromNodeLabels(assignedNode.Labels[corev1.LabelHostname], assignedNode.Labels[corev1.LabelTopologyZone], assignedNode.Labels[corev1.LabelTopologyRegion])
+
+				if nodeLabels != labelsAnnotation {
+					err := kverrors.New("Topology key pod annotation does not match its node's labels")
+					r.Log.Error(err, "Topology key pod annotation does not match its node's labels")
+					return nil
+					// still don't know how to handle it from here
+
+					/*return status.DegradedError{
+						Reason: lokiv1.ReasonZoneLabelsMismatch,
+					}*/
```

Comment:
This function should be limited only for enqueueing the LokiStack resources when it detects label changes. A secondary helper the internal helpers should re-apply the check and then update the degraded status.

Reply:
I see. This part was a bit confusing to me. thanks!